### PR TITLE
Implementing "Add" and "Edit" functionality for user table.

### PR DESCRIPTION
**Pull request by SolomiiaLomnytska22**
**Refs:**
#7 (Implement 'add' and 'edit' user functionality)
**Proposed changes:**
- Added v-model bindings in input components.
- Added data editing functionality in the same input form as "Add" functionality has been implemented (UserInfoModal.vue and UserInfoForm.vue).
- Added button to open editing modal form and function for saving new edited data on JSON-server.

User interface before implementing changes:
Screenshot 2023-11-28 at 18 28 05

User interface after implementing changes:
Screenshot 2023-11-28 at 18 29 07
Screenshot 2023-11-28 at 18 29 54

**Verification:**
To add new user:
1. Run JSON-server using npm run start-json-server.
2. Open User Table page.
3. Click "Add" button on the right side of the page.
4. Enter new user data in a modal pop-up.
5. Click "Add" button in the bottom of the modal.

To edit existing user data:
1. Run JSON-server using npm run start-json-server.
2. Open User Table page.
3. Click "Edit" button in the row with desired user's data.
4. Edit user data in a modal pop-up.
5. Click "Save" button in the bottom of the modal.

Tested locally. All implemented as well as already existing functionality is supported.
User adding process video:
https://github.com/SolomiiaLomnytska22/solomiia-training/assets/136481621/b4f587d6-1d1e-4170-804b-662454d083e2
User editing process video: 
https://github.com/SolomiiaLomnytska22/solomiia-training/assets/136481621/879b63f5-fa9f-4182-888f-e62669bd9651

### CODE REVIEW

File: app/src/components/common/Button.vue [hunk 1]
```diff
@@ -21,7 +21,7 @@ export default class Button extends Vue {
     validator: (value: string) => [ 'primary', 'secondary' ].includes(value)
   })
   styleType!: string
-  @Prop({ required: true }) title!: string
+  /*@Prop({ required: true }) title!: string*/
```
Comment (kaytee-te) [inline app/src/components/common/Button.vue]: Did you leave this commented code on purpose?

File: app/src/components/users/UserInfoForm.vue [hunk 2]
```diff
@@ -1,61 +1,56 @@
...
-      />
-      <Button
-        type="submit"
-        title="Add"
-      />
+      >
+        Close
+      </Button>
+      <Button type="submit">
+        {{ action === 'add' ? 'Add' : 'Save' }}
```
Comment (kaytee-te) [inline app/src/components/users/UserInfoForm.vue]: Let's move this condition to getter.

**Comment by kaytee-te** [comment]
Good job overall!

File: app/src/components/users/UserInfoForm.vue (hunk 2 above)
Comment (SolomiiaLomnytska22) [reply app/src/components/users/UserInfoForm.vue]: Thanks! Created getter in the recent commit.

File: app/src/components/common/Button.vue (hunk 1 above)
Comment (SolomiiaLomnytska22) [reply app/src/components/common/Button.vue]: Thanks for reminding, fixed!

Comment (SolomiiaLomnytska22) [reply app/src/components/common/Button.vue]: Thanks for reminding, fixed!

File: app/src/components/users/UserInfoForm.vue [hunk 3]
```diff
@@ -121,17 +132,27 @@ export default class AddUserForm extends Vue {
       })
   }
 
-  updateUser (field: string, value: string): void {
-    this.$set(this.newUserData, field, value)
+  saveUser (): void {
+    axios
```
Comment (marina-grechukha) [inline app/src/components/users/UserInfoForm.vue]: I suggest using ES7 syntax for handle asynchronous code and replacing .then with async/await

File: app/src/components/users/UserInfoForm.vue [hunk 4]
```diff
@@ -121,17 +132,27 @@ export default class AddUserForm extends Vue {
...
+      .put(
+        `http://localhost:3000/users/${this.newUserData.id}`,
+        this.newUserData
+      )
+      .then((response) => {
+        if (response.status === 200) {
+          this.$emit('user-added')
+          this.closeClick()
+        } else {
+          window.alert('Error while editing user: ' + response.statusText)
```
Comment (marina-grechukha) [inline app/src/components/users/UserInfoForm.vue]: also we can handle it in `catch` block

File: app/src/components/users/UserInfoForm.vue (hunk 3 above)
Comment (marina-grechukha) [reply app/src/components/users/UserInfoForm.vue]: the first article I found with a pretty good explanation of why async/await is better)
https://www.smashingmagazine.com/2020/11/comparison-async-await-versus-then-catch/

File: app/src/components/users/UserInfoModal.vue [hunk 5]
```diff
@@ -9,9 +9,11 @@
...
     >
       <div class="heading">
-        <h2>Add New User</h2>
+        <h2>{{ selectedUser === null ? 'Add New User' : 'Edit User' }}</h2>
       </div>
-      <AddUserForm
+      <UserInfoForm
+        :action="selectedUser === null ? 'add' : 'edit'"
+        :selected-user="selectedUser"
         @close="closeAddUserModal"
```
Comment (marina-grechukha) [inline app/src/components/users/UserInfoModal.vue]:
```suggestion
        @close="closeUserInfoModal"
```

File: app/src/Users.vue [hunk 6]
```diff
@@ -4,22 +4,27 @@
...
-        title="Add"
-        @click="showAddUserModal = true"
-      />
-      <AddUserModal
+        @click="handleAddClick"
+      >
+        Add
+      </Button>
+      <UserInfoModal
         :show-modal="showAddUserModal"
```
Comment (marina-grechukha) [inline app/src/Users.vue]:
```suggestion
        :show-modal="showUserInfoModal"
```

File: app/src/Users.vue [hunk 7]
```diff
@@ -28,19 +33,30 @@ import { User } from './types'
...
 export default class Users extends Vue {
   users: User[] = []
   showAddUserModal = false
+  selectedUser: User | null = null
 
   mounted () {
     this.getData()
   }
 
+  handleEditEvent (user: User) {
```
Comment (marina-grechukha) [inline app/src/Users.vue]:
```suggestion
  handleEditUser (user: User) {
```

File: app/src/Users.vue [hunk 8]
```diff
@@ -28,19 +33,30 @@ import { User } from './types'
...
+    this.selectedUser = user
+    this.showAddUserModal = true
+  }
+
+  handleAddClick (): void {
```
Comment (marina-grechukha) [inline app/src/Users.vue]:
```suggestion
  handleAddUser (): void {
```

File: app/src/components/users/UserInfoForm.vue (hunk 3 above)
Comment (SolomiiaLomnytska22) [reply app/src/components/users/UserInfoForm.vue]: Thanks, tried to fix it in the recent commit.

**Review by marina-grechukha — APPROVED** [review]
looks good 👍  In continuation of this work, I suggest adding toasts to display success and errors. 
I created a new issue for this https://github.com/SolomiiaLomnytska22/solomiia-training/issues/12

**Review by kaytee-te — APPROVED** [review]

**Pull request merged**